src/Components/ServiceScene/LogsJsonScene.tsx

```diff
@@ -143,13 +149,254 @@ export class LogsJsonScene extends SceneObjectBase<LogsJsonSceneState> {
+      (visible: boolean) => {
+        model.setState({ showMetadata: visible });
+        setJsonMetadataVisibility(visible);
+      },
+      [model]
+    );
+
+    const onToggleLabelsClick = useCallback(
+      (visible: boolean) => {
+        model.setState({ showLabels: visible });
+        setJsonLabelsVisibility(visible);
+      },
+      [model]
+    );
+
+    const onToggleHighlightClick = useCallback(
+      (visible: boolean) => {
+        model.setState({ showHighlight: visible });
+        setJsonHighlightVisibility(visible);
+      },
+      [model]
+    );
+
+    return (
+      // @ts-expect-error todo: fix this when https://github.com/grafana/grafana/issues/103486 is done
+      <PanelChrome
+        padding={'none'}
+        showMenuAlways={true}
+        statusMessage={$data.state.data?.errors?.[0].message}
+        loadingState={$data.state.data?.state}
+        title={
+          <>
+            JSON <Badge color={'blue'} text={'Experimental'} />
+          </>
+        }
+        menu={menu ? <menu.Component model={menu} /> : undefined}
+        actions={<LogsPanelHeaderActions vizType={visualizationType} onChange={logsListScene.setVisualizationType} />}
+      >
+        <div className={styles.container}>
+          {lineField?.values && lineField?.values.length > 0 && (
+            <LogListControls
+              showHighlight={showHighlight}
+              onToggleHighlightClick={onToggleHighlightClick}
+              showMetadata={showMetadata}
+              onToggleStructuredMetadataClick={onToggleStructuredMetadataClick}
+              showLabels={showLabels}
+              onToggleLabelsClick={onToggleLabelsClick}
+              sortOrder={sortOrder}
+              onSortOrderChange={model.handleSortChange}
+              onScrollToBottomClick={onScrollToBottomClick}
+              onScrollToTopClick={onScrollToTopClick}
+            />
+          )}
+          {dataFrame && lineField?.values && lineField?.values.length > 0 && (
+            <div className={styles.JSONTreeWrap} ref={scrollRef}>
+              {jsonFiltersSupported === false && (
+                <Alert severity={'warning'} title={'JSON filtering requires Loki 3.5.0.'}>
+                  This view will be read only until Loki is upgraded to 3.5.0
+                </Alert>
+              )}
+              {lineField.values.length > 0 && hasJsonFields === false && (
+                <>
+                  <Alert severity={'info'} title={'No JSON fields detected'}>
+                    This view is built for JSON log lines, but none were detected. Switch to the Logs or Table view for
+                    a better experience.
+                  </Alert>
+                </>
+              )}
+              <JSONTree
+                data={lineField.values}
+                hideRootExpand={true}
+                valueWrap={''}
+                getItemString={(_, data, itemType, itemString, keyPath) => {
```

Comment:
Extracting these functions would be great and potentially aid with performance. Every inline function will be a potential performance regression.

Reply:
They've been extracted into new components in new files, should be a bit easier to work with now.